main.go

```diff
@@ -14,115 +14,247 @@
+	endpoint := sriovns.NewServer(
+		ctx,
 		config.Name,
 		authorize.NewServer(),
 		spiffejwt.TokenGeneratorFunc(source, config.MaxTokenLifetime),
+		pciPool,
+		resourcePool,
+		sriovConfig,
+		config.VFIOPath, config.CgroupPath,
 		&config.ConnectTo,
-		grpc.WithTransportCredentials(credentials.NewTLS(tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny()))),
+		grpc.WithTransportCredentials(
+			grpcfd.TransportCredentials(
+				credentials.NewTLS(tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny())),
+			),
+		),
 		grpc.WithDefaultCallOptions(grpc.WaitForReady(true)),
 	)
 
-	// Create GRPC Server
-	// TODO - add ServerOptions for Tracing
-	server := grpc.NewServer(grpc.Creds(credentials.NewTLS(tlsconfig.MTLSServerConfig(source, source, tlsconfig.AuthorizeAny()))))
+	// ********************************************************************************
+	log.Entry(ctx).Infof("executing phase 7: create grpc server and register sriovns (time since start: %s)", time.Since(starttime))
+	// ********************************************************************************
+	tmpDir, err := ioutil.TempDir("", "sriov-forwarder")
+	if err != nil {
+		log.Entry(ctx).Fatalf("error creating tmpDir: %+v", err)
+	}
+	defer func() { _ = os.RemoveAll(tmpDir) }()
+	listenOn := &url.URL{Scheme: "unix", Path: path.Join(tmpDir, "listen_on.io.sock")}
+
+	server := grpc.NewServer(append(
+		spanhelper.WithTracing(),
+		grpc.Creds(
+			grpcfd.TransportCredentials(
+				credentials.NewTLS(tlsconfig.MTLSServerConfig(source, source, tlsconfig.AuthorizeAny())),
+			),
+		))...)
 	endpoint.Register(server)
-	srvErrCh := grpcutils.ListenAndServe(ctx, &config.ListenOn, server)
+	srvErrCh := grpcutils.ListenAndServe(ctx, listenOn, server)
 	exitOnErr(ctx, cancel, srvErrCh)
+
+	// ********************************************************************************
+	log.Entry(ctx).Infof("executing phase 8: register %s with the registry (time since start: %s)", config.NSName, time.Since(starttime))
+	// ********************************************************************************
+	registryCreds := credentials.NewTLS(tlsconfig.MTLSClientConfig(source, source, tlsconfig.AuthorizeAny()))
+	registryCreds = grpcfd.TransportCredentials(registryCreds)
+	registryCC, err := grpc.DialContext(ctx,
+		grpcutils.URLToTarget(&config.ConnectTo),
+		append(
+			spanhelper.WithTracingDial(),
+			grpc.WithTransportCredentials(registryCreds),
+			grpc.WithBlock(),
+		)...,
+	)
+	if err != nil {
+		log.Entry(ctx).Fatalf("failed to connect to registry: %+v", err)
+	}
+
+	registryClient := registrychain.NewNetworkServiceEndpointRegistryClient(
+		// TODO - add refresh
+		registrysendfd.NewNetworkServiceEndpointRegistryClient(),
+		registryapi.NewNetworkServiceEndpointRegistryClient(registryCC),
+	)
+	// TODO - something smarter for expireTime
+	expireTime, err := ptypes.TimestampProto(time.Now().Add(config.MaxTokenLifetime))
```

Comment:
I think @denis-tingajkin already had some chain element for registry, to perform this automatically and it should include automatic re-register to prevent registration expire.

Reply:
I feel `refresh.NewNetworkServiceEndpointRegistryClient()` should be fine here.

Reply:
Yep, fixed.